wgcfg/ip.go

```diff
@@ -0,0 +1,152 @@
+package wgcfg
+
+import (
+	"fmt"
+	"math"
+	"net"
+)
+
+// IP is an IPv4 or an IPv6 address.
+//
+// Internally the address is always represented in its IPv6 form.
+// IPv4 addresses use the IPv4-in-IPv6 syntax.
+type IP struct {
+	Addr [16]byte
+}
+
+func (ip IP) String() string { return net.IP(ip.Addr[:]).String() }
+
+// IP converts ip into a standard library net.IP.
+func (ip IP) IP() net.IP { return net.IP(ip.Addr[:]) }
+
+// Is6 reports whether ip is an IPv6 address.
+func (ip IP) Is6() bool { return !ip.Is4() }
+
+// Is4 reports whether ip is an IPv4 address.
+func (ip IP) Is4() bool {
```

Comment:
Typically throughout WireGuard, we've used `[4]byte` for v4 and `[16]byte` for v6, considering the Go standard library's choice of v6-mapped-v4 to be a mistake. Especially since somebody might legitimately put a ::ffff:1.2.3.4 address through WireGuard and expect it to be v6 and not punned into v4.

Reply:
Using the newer [net/netip](https://pkg.go.dev/net/netip) from the Go standard library would resolve the v6-mapped-v4 issue. Seems like Go wanted to improve it too.